metals/src/main/scala/scala/meta/internal/metals/scalacli/ScalaCli.scala

```diff
@@ -60,84 +59,93 @@ class ScalaCli(
+      out
+    }(Seq.empty)
   }
 
   val buildTargetsData = new TargetData
+
   def lastImportedBuild: ImportedBuild =
-    lastImportedBuild0
+    ifConnectedOrElse(_.importedBuild)(ImportedBuild.empty)
+
   def buildServer: Option[BuildServerConnection] =
-    buildServer0
-  def roots: Seq[AbsolutePath] =
-    roots0
+    ifConnectedOrElse(conn => Option(conn.connection))(None)
 
   def importBuild(): Future[Unit] =
-    buildServer0 match {
-      case None =>
-        Future.failed(new Exception("No Scala CLI server running"))
-      case Some(conn) =>
-        compilers().cancel()
-
-        for {
-          build0 <- statusBar().trackFuture(
-            "Importing Scala CLI sources",
-            ImportedBuild.fromConnection(conn),
-          )
-          _ = {
-            lastImportedBuild0 = build0
+    ifConnectedOrElse { st =>
+      compilers().cancel()
+
+      statusBar()
+        .trackFuture(
+          "Importing Scala CLI sources",
+          ImportedBuild.fromConnection(st.connection),
+        )
+        .flatMap { build0 =>
+          if (state.compareAndSet(st, st.copy(importedBuild = build0))) {
             val targets = build0.workspaceBuildTargets.getTargets.asScala
             val connections =
-              targets.iterator.map(_.getId).map((_, conn)).toList
+              targets.iterator.map(_.getId).map((_, st.connection)).toList
             buildTargetsData.resetConnections(connections)
-          }
-          _ <- indexWorkspace()
-          allSources0 = allSources()
-          toCompile = buffers.open.toSeq.filter(p =>
-            allSources0.exists(root =>
-              root == p || p.toNIO.startsWith(root.toNIO)
-            )
-          )
-          _ <- Future.sequence(
-            compilations
-              .cascadeCompileFiles(toCompile) ::
-              compilers().load(toCompile) ::
-              Nil
-          )
-        } yield ()
-    }
+
+            for {
+              _ <- indexWorkspace()
+              allSources0 = allSources()
+              toCompile = buffers.open.toSeq.filter(p =>
+                allSources0
+                  .exists(root => root == p || p.toNIO.startsWith(root.toNIO))
+              )
+              _ <- Future.sequence(
+                compilations
+                  .cascadeCompileFiles(toCompile) ::
+                  compilers().load(toCompile) ::
+                  Nil
+              )
+            } yield ()
+          } else importBuild()
+        }
+    }(Future.failed(new Exception("No Scala CLI server running")))
 
   private def disconnectOldBuildServer(): Future[Unit] = {
-    if (buildServer0.isDefined)
-      scribe.info("disconnected: Scala CLI server")
-    buildServer0 match {
-      case None => Future.unit
-      case Some(value) =>
-        val allSources0 = allSources()
-        buildServer0 = None
-        lastImportedBuild0 = ImportedBuild.empty
-        roots0 = Nil
-        cancelables.cancel()
```

Comment:
[note]
cancellation is done in shutdown
https://github.com/scalameta/metals/blob/f48987e43d3b36c530da396ba94290d25e3332f9/metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala#L122